lib/active_model/serializer/adapter/json.rb

```diff
@@ -4,44 +4,130 @@ class ActiveModel::Serializer::Adapter::Json < ActiveModel::Serializer::Adapter
+
+          included_associations.each do |association|
+            serializer = association.serializer
+            opts = association.options
+            key = association.key
+
+            # sanity check if the association has nesting data
+            has_nesting = nested_associations[key].present?
+            if has_nesting
+              include_options_from_parent = { include: nested_associations[key] }
+              opts = opts.merge(include_options_from_parent)
             end
 
-            serializer.associations.each do |association|
-              serializer = association.serializer
-              opts = association.options
-
-              if serializer.respond_to?(:each)
-                array_serializer = serializer
-                hash[association.key] = array_serializer.map do |item|
-                  cache_check(item) do
-                    item.attributes(opts)
-                  end
-                end
-              else
-                hash[association.key] =
-                  if serializer && serializer.object
-                    cache_check(serializer) do
-                      serializer.attributes(options)
-                    end
-                  elsif opts[:virtual_value]
-                    opts[:virtual_value]
-                  end
-              end
+            if serializer.respond_to?(:each)
+              parent[key] = to_many_relationships_for(serializer, opts)
+            else
+              parent[key] = to_one_relationship_for(serializer, opts)
             end
-            result = core.merge hash
           end
 
-          { root => result }
+          parent
         end
 
-        def fragment_cache(cached_hash, non_cached_hash)
-          ActiveModel::Serializer::Adapter::Json::FragmentCache.new().fragment_cache(cached_hash, non_cached_hash)
+        def relevant_included_associations_for(serializer, nested_associations = {})
+          if nested_associations.present?
+            serializer.associations.select do |association|
+              # nested_associations is a hash of:
+              #   key => nested association to include
+              nested_associations.key?(association.name)
+            end
+          else
+            serializer.associations
+          end
+        end
+
+        # add a singular relationship
+        # the options should always belong to the serializer
+        def to_one_relationship_for(serializer, options)
+          serialized_relationship = serialized_or_virtual_of(serializer, options)
+
+          nested_associations_to_include = options[:include]
+          if nested_associations_to_include.present?
+            serialized_relationship = add_resource_relationships(
+              serialized_relationship,
+              serializer,
+              nested_associations_to_include)
+          end
+
+          serialized_relationship
+        end
+
+        # add a many relationship
+        def to_many_relationships_for(serializer, options)
```

Comment:
Indeed, there is no textual distinction in the rendered json, only that `to-one`s are represented by `null` or a hash, and `to-many`s are represented by a (possibly empty) array.

I do not agree that there should be a plural next to `many` (I forgot the grammatical terminology to talk about that, but here `to_many` is just qualifying a single relationship). 

Thing is, in JsonApi, you never talk about "what's on the other end of the relationship", you only talk about relationships, and resources (those that are on the other end of a relationship just happen to be in the `included` part of the document), which explains why those questions did not really arise in the JsonApi adapter.